optuna/study/study.py

```diff
@@ -760,8 +746,7 @@ def objective(trial):
 
         """
 
-        if self._optimize_lock.acquire(False):
-            self._optimize_lock.release()
+        if not _in_optimize_loop.get():
```

Comment:
Do you think we need to keep this thread local context, again, that may still span multiple studies? 
(That is, you can still call `Study.stop` on an "inner study", even when it's not running `Study.optimize`, which may be non-intuitive).

What would happen by the way if we keep a local member variable bound to each study instead? Or, I think we actually can discuss this error too, whether we need to keep it. Do you see negative consequences in removing it?

Reply:
Good point.
I think we should avoid removing it, at least in this PR, since it may be non-intuitive as you said. "keeping a local member variable bound to each study" sounds reasonable option for us. I've just opened #4068 that just replacing `threading.Lock` with `threading.local`, so that the above concern can be resolved.